src/libaktualizr/primary/eventsinterpreter.cc

```diff
@@ -19,56 +19,117 @@ EventsInterpreter::~EventsInterpreter() {
 
-    if (event->variant == "SendDeviceDataComplete") {
-      *commands_channel << std::make_shared<command::FetchMeta>();
-    } else if (event->variant == "FetchMetaComplete") {
-      *commands_channel << std::make_shared<command::CheckUpdates>();
-    } else if (event->variant == "UpdateAvailable") {
-      if (config.uptane.running_mode == RunningMode::kInstall) {
-        *commands_channel << std::make_shared<command::UptaneInstall>(
-            dynamic_cast<event::UpdateAvailable*>(event.get())->updates);
-      } else if (config.uptane.running_mode != RunningMode::kCheck) {
-        *commands_channel << std::make_shared<command::StartDownload>(
-            dynamic_cast<event::UpdateAvailable*>(event.get())->updates);
-      } else {
-        *commands_channel << std::make_shared<command::Shutdown>();
-      }
-    } else if (event->variant == "DownloadComplete") {
-      if (config.uptane.running_mode == RunningMode::kFull || config.uptane.running_mode == RunningMode::kOnce) {
-        *commands_channel << std::make_shared<command::UptaneInstall>(
-            dynamic_cast<event::DownloadComplete*>(event.get())->updates);
-      } else {
-        *commands_channel << std::make_shared<command::Shutdown>();
-      }
-    } else if (event->variant == "InstallComplete") {
-      if (config.uptane.running_mode == RunningMode::kFull || config.uptane.running_mode == RunningMode::kOnce) {
-        *commands_channel << std::make_shared<command::PutManifest>();
-      } else {
-        *commands_channel << std::make_shared<command::Shutdown>();
-      }
+    std::shared_ptr<command::BaseCommand> next_command;
+    switch (running_mode) {
+      case RunningMode::kFull:
+        next_command = handle_cycle(*event, true);
+        break;
+      case RunningMode::kOnce:
+        next_command = handle_cycle(*event, false);
+        break;
+      case RunningMode::kCheck:
+        next_command = handle_check(*event);
+        break;
+      case RunningMode::kDownload:
+        next_command = handle_download(*event);
+        break;
+      case RunningMode::kInstall:
+        next_command = handle_install(*event);
+        break;
+      default:
+        LOG_ERROR << "Unknown running mode " << StringFromRunningMode(running_mode);
     }
-    if (event->variant == "UptaneTimestampUpdated" || event->variant == "PutManifestComplete" ||
-        event->variant == "Error") {
-      // These events indicates the end of pooling cycle
-      if (config.uptane.running_mode == RunningMode::kFull) {
-        std::this_thread::sleep_for(std::chrono::seconds(config.uptane.polling_sec));
-        *commands_channel << std::make_shared<command::FetchMeta>();
-      } else {
-        *commands_channel << std::make_shared<command::Shutdown>();
-      }
+
+    if (next_command == nullptr) {
+      LOG_ERROR << "No command to run after event " << event->variant << " in mode "
+                << StringFromRunningMode(running_mode);
```

Comment:
I'm not sure this is really an error. In my PR (https://github.com/advancedtelematic/aktualizr/pull/871) I've created a `kNone` running mode to indicate completely manual control, i.e. you can freely add commands to the channel, and events are produced as a result. Thus it is valid that not every event is followed by another command.

Reply:
Should it be a warning or just stay silent?

Reply:
Actually, we could leave it there and only silence it if `RunningMode.kNone` (which I will do if we merge this first). In my new situation, it is not an error, but for the existing modes it is.